### PR TITLE
Added max_s3_upload_threads_per_task option to Redshift

### PR DESCRIPTION
**Pull request by t3t5u**
The changes are essentially the same as in [this PR](https://github.com/embulk/embulk-output-jdbc/pull/311).
I've fixed the points commented on in the above review.

### CODE REVIEW

**Comment by t3t5u** [comment]
@dmikurube @hiroyuki-sato 
Please review. 🙇‍♂️

**Review by hiroyuki-sato — COMMENTED** [review]
I just want to know why this PR changed some part (i.e default value) from the original.
I don't have any knowledge which is better yet.

File: embulk-output-redshift/src/main/java/org/embulk/output/RedshiftOutputPlugin.java [hunk 1]
```diff
@@ -83,6 +83,10 @@ public interface RedshiftPluginTask extends AwsCredentialsTaskWithPrefix, Plugin
         @ConfigDefault("true")
         public boolean getDeleteS3TempFile();
 
+        @Config("max_s3_upload_threads_per_task")
+        @ConfigDefault("null")
```
Comment (hiroyuki-sato) [inline embulk-output-redshift/src/main/java/org/embulk/output/RedshiftOutputPlugin.java:87]: The default value of the original PR is `5`. Why you change this value?
https://github.com/embulk/embulk-output-jdbc/pull/311/files#diff-3ad638974347d49a3df0afc40dc2fa00382b91105aa42eeec25468daccbfc434R54

Any update after this comment in trocco servcie?
https://github.com/embulk/embulk-output-jdbc/pull/311

> As newCachedThreadPool would create new thread for execution if and only if current thread are all occupied, it seems tenable to pass a maximum of thread to newCachedThreadPool. However, due to the execution mechanism of embulk is ansync we may have troubles in how to determine the max of it. The most simple way to resolve is just make the maximum of each thread pool as a argument. After some performance test, it seems fine to make the default value of thread_maximum as 5.

File: embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java [hunk 2]
```diff
@@ -82,7 +83,9 @@ public RedshiftCopyBatchInsert(JdbcOutputConnector connector,
         this.credentialsProvider = credentialsProvider;
         this.s3 = new AmazonS3Client(credentialsProvider);  // TODO options
         this.sts = new AWSSecurityTokenServiceClient(credentialsProvider);  // options
-        this.executorService = Executors.newCachedThreadPool();
+        this.executorService = maxS3UploadThreadsPerTask != null
+                ? Executors.newFixedThreadPool(maxS3UploadThreadsPerTask)
+                : Executors.newCachedThreadPool();
         this.uploadAndCopyFutures = new ArrayList<Future<Void>>();
```
Comment (hiroyuki-sato) [inline embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java:89]: The original PR change thread pool from `newThreadPool` to `Executors.newFixedThreadPool(threadMaximum)`. 
Why you change this part from the origianl?

File: embulk-output-redshift/src/main/java/org/embulk/output/RedshiftOutputPlugin.java (hunk 1 above)
Comment (t3t5u) [reply embulk-output-redshift/src/main/java/org/embulk/output/RedshiftOutputPlugin.java:87]: I've kept the original behavior by default, due to the requested in the following comment.

https://github.com/embulk/embulk-output-jdbc/pull/311#pullrequestreview-1241537103
```
It changes the thread behavior even if thread_maximum is not specified.
newCachedThreadPool() and newFixedThreadPool(5) behave different.
Can you make the it Optional<Interger>, and keep the behavior when the option is not specified?
```

> Any update after this comment in trocco servcie?

No, the maximum number of threads remains `5` in the trocco.

File: embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java (hunk 2 above)
Comment (t3t5u) [reply embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java:89]: For the same reason as above, if `max_s3_upload_threads_per_task` is not specified, continue to use `newCachedThreadPool` to keep the original behavior.

**Review by dmikurube — APPROVED** [review]
This change basically looks good to me, as it keeps the original behavior, and it follows the feedbacks in the previous pull request.

Two points :
1. Wanted to hear other comments from @embulk/jdbc-maintainers
2. If some more change requests are expected to come from trocco, we may want to consider re-organizing the maintainer team and the maintenance policies. We now have some discussions on Zulip with @kekekenta, and let's keep going on.

File: embulk-output-redshift/src/main/java/org/embulk/output/RedshiftOutputPlugin.java (hunk 1 above)
Comment (hiroyuki-sato) [reply embulk-output-redshift/src/main/java/org/embulk/output/RedshiftOutputPlugin.java:87]: OK, Thanks. Do you think whether is is better to change `5` by default in future release?

File: embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java (hunk 2 above)
Comment (hiroyuki-sato) [reply embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java:89]: OK. Thanks

**Review by hiroyuki-sato — COMMENTED** [review]
Hello, @t3t5u Thank you for your comment.

What is the relationship #311 and this PR?
IIUC, Both PRs propose from trocco team.
I tought This PR was proposed for replace #311. Correct?

**Review by hiroyuki-sato — APPROVED** [review]
Basically, This PR LGTM. If it is better to change `max_s3_upload_threads_per_task` default value, let's change default in future release. For preparation, adding log message `this value will change in future release...` is one of choice.

otherwise, Let's keep current behavior.

File: embulk-output-redshift/src/main/java/org/embulk/output/RedshiftOutputPlugin.java (hunk 1 above)
Comment (t3t5u) [reply embulk-output-redshift/src/main/java/org/embulk/output/RedshiftOutputPlugin.java:87]: I don't think there is a problem with the change, but I don't think it needs to be changed.
I think users would change the configuration as needed.

**Comment by t3t5u** [comment]
@hiroyuki-sato 

> I tought This PR was proposed for replace #311. Correct?

That is correct. This PR is a reworked replacement for #311.

> otherwise, Let's keep current behavior.

Thank you for the advice.
I think the default behavior doesn't need to be changed at this time.

**Comment by hiroyuki-sato** [comment]
@t3t5u Thanks! LGTM👍

**Comment by dmikurube** [comment]
Thanks for your contribution. Merging this for v0.10.5...

**Pull request merged**